src/librustc_trait_selection/traits/fulfill.rs

```diff
@@ -318,267 +319,223 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
         let infcx = self.selcx.infcx();
 
         match obligation.predicate.kind() {
-            ty::PredicateKind::Trait(ref data, _) => {
-                let trait_obligation = obligation.with(*data);
-
-                if obligation.predicate.is_global() {
-                    // no type variables present, can use evaluation for better caching.
-                    // FIXME: consider caching errors too.
-                    if infcx.predicate_must_hold_considering_regions(&obligation) {
-                        debug!(
-                            "selecting trait `{:?}` at depth {} evaluated to holds",
-                            data, obligation.recursion_depth
-                        );
-                        return ProcessResult::Changed(vec![]);
+            ty::PredicateKind::ForAll(binder) => match binder.skip_binder().kind() {
+                ty::PredicateKind::ForAll(_) => bug!("unexpected forall"),
+                // Evaluation will discard candidates using the leak check.
+                // This means we need to pass it the bound version of our
+                // predicate.
+                &ty::PredicateKind::Atom(atom) => match atom {
+                    ty::PredicateAtom::Trait(trait_ref, _constness) => {
+                        let trait_obligation = obligation.with(Binder::bind(trait_ref));
+
+                        self.process_trait_obligation(
+                            obligation,
+                            trait_obligation,
+                            &mut pending_obligation.stalled_on,
+                        )
                     }
-                }
+                    ty::PredicateAtom::Projection(projection) => {
```

Comment:
I think that Projections no longer need special casing (they no longer run the leak check at the top level).

Reply:
Doesn't seem like it.

Removing the specialcase for projections results in 
```
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling compiler_builtins v0.1.32
   Compiling core v0.0.0 (/home/lcnr/rust2/src/libcore)
   Compiling cc v1.0.57
   Compiling libc v0.2.71
   Compiling autocfg v0.1.7
   Compiling std v0.0.0 (/home/lcnr/rust2/src/libstd)
   Compiling hashbrown v0.6.2
   Compiling backtrace-sys v0.1.37
   Compiling unwind v0.0.0 (/home/lcnr/rust2/src/libunwind)
error[E0599]: no method named `count` found for struct `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>` in the current scope
    --> src/libcore/num/bignum.rs:173:74
     |
100  | / macro_rules! define_bignum {
101  | |     ($name:ident: type=$ty:ty, n=$n:expr) => {
102  | |         /// Stack-allocated arbitrary-precision (up to certain limit) integer.
103  | |         ///
...    |
173  | |                 let zeros = digits.iter().rev().take_while(|&&x| x == 0).count();
     | |                                                            ------------  ^^^^^ method not found in `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>`
     | |                                                            |
     | |                                                            doesn't satisfy `<_ as ops::function::FnOnce<(&&u32,)>>::Output = bool`
     | |                                                            doesn't satisfy `_: ops::function::FnMut<(&&u32,)>`
...    |
477  | |     };
478  | | }
     | |_- in this expansion of `define_bignum!`
...
483  |   define_bignum!(Big32x40: type=Digit32, n=40);
     |   --------------------------------------------- in this macro invocation
     | 
    ::: src/libcore/iter/adapters/mod.rs:1803:1
     |
1803 |   pub struct TakeWhile<I, P> {
     |   --------------------------
     |   |
     |   method `count` not found for this
     |   doesn't satisfy `_: iter::traits::iterator::Iterator`
     |
     = note: the method `count` exists but the following trait bounds were not satisfied:
             `<[closure@src/libcore/num/bignum.rs:173:60: 173:72] as ops::function::FnOnce<(&&u32,)>>::Output = bool`
             which is required by `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             `[closure@src/libcore/num/bignum.rs:173:60: 173:72]: ops::function::FnMut<(&&u32,)>`
             which is required by `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             which is required by `&mut iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u32>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`

error[E0599]: no method named `count` found for struct `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>` in the current scope
    --> src/libcore/num/bignum.rs:173:74
     |
100  | / macro_rules! define_bignum {
101  | |     ($name:ident: type=$ty:ty, n=$n:expr) => {
102  | |         /// Stack-allocated arbitrary-precision (up to certain limit) integer.
103  | |         ///
...    |
173  | |                 let zeros = digits.iter().rev().take_while(|&&x| x == 0).count();
     | |                                                            ------------  ^^^^^ method not found in `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>`
     | |                                                            |
     | |                                                            doesn't satisfy `<_ as ops::function::FnOnce<(&&u8,)>>::Output = bool`
     | |                                                            doesn't satisfy `_: ops::function::FnMut<(&&u8,)>`
...    |
477  | |     };
478  | | }
     | |_- in this expansion of `define_bignum!`
...
488  |       define_bignum!(Big8x3: type=u8, n=3);
     |       ------------------------------------- in this macro invocation
     | 
    ::: src/libcore/iter/adapters/mod.rs:1803:1
     |
1803 |   pub struct TakeWhile<I, P> {
     |   --------------------------
     |   |
     |   method `count` not found for this
     |   doesn't satisfy `_: iter::traits::iterator::Iterator`
     |
     = note: the method `count` exists but the following trait bounds were not satisfied:
             `<[closure@src/libcore/num/bignum.rs:173:60: 173:72] as ops::function::FnOnce<(&&u8,)>>::Output = bool`
             which is required by `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             `[closure@src/libcore/num/bignum.rs:173:60: 173:72]: ops::function::FnMut<(&&u8,)>`
             which is required by `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             `iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`
             which is required by `&mut iter::adapters::TakeWhile<iter::adapters::Rev<slice::Iter<'_, u8>>, [closure@src/libcore/num/bignum.rs:173:60: 173:72]>: iter::traits::iterator::Iterator`

error: internal compiler error: src/librustc_middle/ich/impls_ty.rs:94:17: StableHasher: unexpected region RePlaceholder(Placeholder { universe: U1, name: BrAnon(0) })
```